### ISSUE TITLE
Merge grid sources into one source

### ISSUE THREAD

**Issue by tischi** [mobie/mobie-viewer-fiji#398]
@constantinpape @K-Meech 

I am planning to look into merging many sources from a grid view into one source on the Java side. The hope is that this would improve the performance of the visualization if there are many (>100) sources in a grid.

I think, in the MoBIE UI this would make no difference, because all the display settings are anyway joined for all the sources in a grid view display.

On the BDV UI this would look different as the sources would not be accessible individually anymore, but just as one source, like in the MoBIE UI.

I am thinking of adding a field `merge` (maybe even default `true`) to the grid view transform and a field `mergedSourceName`.

Implementation wise I think it would be something like

```
RandomAccessibleInterval< T > cachedCellImg = new ReadOnlyCachedCellImgFactory().create(
				gridBoundaryDimensions, // e.g. the well or the plate
				T, // data type
				loader, // loads from all the sources in the grid
				ReadOnlyCachedCellImgOptions.options().cellDimensions( gridCellDimensions ) );
```

for each `RandomAccessibleInterval` for each resolution level in the (to be developed) `GridSource`.

Sounds good?

**Comment by constantinpape** [comment]
> I am thinking of adding a field `merge` (maybe even default `true`) to the grid view transform and a field `mergedSourceName`.

Do you want to add this to the spec as well? It feels a bit like a technical detail, but if you want we can add it.
But maybe it's also ok to just merge the grid sources by default?!

> Sounds good?

Yes!

**Comment by tischi** [comment]
Maybe some people want to manually move around individual sources in a grid view, I would thus make it optional.

**Comment by constantinpape** [comment]
> Maybe some people want to manually move around individual sources in a grid view, I would thus make it optional.

Ok, that's a good point.
Then let's see how you succeed in merging the sources and then we can think about how to add this to the spec.

**Comment by tischi** [comment]
> Then let's see how you succeed in merging the sources 

Yeah, looking forward to some serious imglib2 coding 😄 probably will have some time tomorrow for this.

**Comment by tischi** [comment]
@constantinpape I disabled the multi-box overlay (the source overview in the top left corner of BDV). I feel that makes it faster to interact. If you happen to be at EMBL, could you try? How is it now compared with the plateviewer? I guess still less responsive?

**Comment by constantinpape** [comment]
I will check it out tomorrow. Though I will not be at EMBL, so I can only properly test it next week.

**Comment by tischi** [comment]
On the other hand also good to know how fast it is from home, because, as far as I remember, the plateViewer also worked OK from home, didn't it?

Another thing: There is an issue with the `GridSource` idea: according to our current spec the `sources = Map< gridId, List sourcenames >` can contain sources of multiple channels that would be all transformed together. A BDV Source however can only contain one color. Maybe we need an entirely new JSON field? Something like `SingleChannelGridSourceTransformerAndMerger`?

Anyway, to get me going, do you know whether we have a project somewhere where we do a grid transform only on a single channel? I guess the covid tomograms?!

**Comment by tischi** [comment]
...I tried opening https://github.com/mobie/covid-tomo-datasets but got errors, maybe we don't have this yet with the current spec?

**Comment by tischi** [comment]
I think we should maybe make a new transform: 

```
gridMerge
- List< Sources >
- List< Positions > // optional
- mergedGridSourceName
```

Where there is simply one source per position.

This could be used to merge site sources into one well => one "well source"; and then for creating the plate view one can use the same logic, but refer to the "well sources".

I think this would also work for the source annotation views as one can refer to, e.g. the created well sources for the well table. 

If we only support this the source annotation views would also become easier because one would simply refer to one source per row and the annotation_id could be replaced by source.

So overall I have the impression that would make everything easier (and hopefully also more performant). 
One drawback would be that one would not be able to address individual sources anymore and e.g. move them around. Maybe something else that I am missing?

**Comment by tischi** [comment]
To get started I would develop this in parallel.
I think two bookmarks for the HTM data would be good

- bm1: just all the nuclei (no tables) the way we did it until now
- bm2: just all the nuclei using a nested gridMerge transform

Then I can develop the gridMerge transform and we can see if it really is more responsive before we change everything else.

**Comment by constantinpape** [comment]
TODO for me: bookmarks with:
- full plate w/ new (merged) + old spec, no tables (= 2 bookmarks)
- 2 wells w/ new + old spec, no tables (= 2 bookmarks)
- make one of the small bms default

**Comment by constantinpape** [comment]
@tischi I updated the covid-if project and we have the 4 new bookmarks now:
- `default`: the 2 wells with the normal grid trafo
- `default_merged`: the 2 wells with merged grid trafo
- `full`: all wells with normal grid trafo
- `full_merged`: all wells with merged grid trafo

I called the merged trafo "gridMerged".

**Comment by tischi** [comment]
@constantinpape Could you please add bookmarks with mergedGrid for **only the nuclei**? It's just easier to debug for me. Thanks!

**Comment by tischi** [comment]
It is working and _much_ more responsive 🥳 
<img width="564" alt="image" src="https://user-images.githubusercontent.com/2157566/128847090-fc4f5448-4640-465a-9f58-f1fd290f3e01.png">

**Comment by constantinpape** [comment]
> It is working and _much_ more responsive partying_face

Awesome! Do you still need the view with only nuclei?

Also let me know when you have updated MoBIE beta, then I will give it a try.

**Comment by tischi** [comment]
> Do you still need the view with only nuclei?

No, it takes a while initially to load everything but then it is super responsive.

> Also let me know when you have updated MoBIE beta, then I will give it a try.

yes, I will...

**Comment by tischi** [comment]
Ok, updated, please try!

**Comment by constantinpape** [comment]
Awesome, this is indeed much much better. It feels even smoother than the plateviewer (after the inital loading).

One minor thing I noticed is that there are some jumps when switching sources on and off (but that's not a big deal, just wanted to keep track of this):

https://user-images.githubusercontent.com/4263537/128867260-be058252-0c0e-485d-adab-45a30a7da8f5.mp4

**Comment by tischi** [comment]
Ah, maybe this is again this issue that different resolution layers are not exactly on top of each other...
I remember something like this in the plateviewer as well. ..

**Comment by constantinpape** [comment]
Ok, quick discussion on how to proceed:
- we rename "mergedGrid -> grid", "grid -> transformGrid"; maybe we completely remove "transformGrid" in the future; but for now we leave it in
- I will update the current project and add views with site and well table (`..._with_table`) and views with segmentations (`..._with_segmentation`)

**Comment by constantinpape** [comment]
@tischi: I think our new grid transform also needs the `centerAtOrigin` field

**Comment by constantinpape** [comment]
and `timepoints` ;)

**Comment by tischi** [comment]
> I think our new grid transform also needs the centerAtOrigin field

Is does have it already

> and timepoints ;)

Yes, I omitted this until now...

**Comment by constantinpape** [comment]
@tischi I updated the project now and there are 8 grid views for all possible combinations of "small" vs. "full", "grid" vs. "transformed grid", "no image tables" vs. "image tables", "no segmentations" vs. "segmentations".
I hope that the view names are telling enough.
Let me know if you run into any issues.
